src/main/kotlin/org/rust/ide/inspections/RsNeedlessLifetimesInspection.kt

```diff
@@ -0,0 +1,246 @@
+    bounds.map { it.name.toReferenceLifetime() }.filterIsInstance<ReferenceLifetime.Named>()
+
+private fun collectLifetimesFromFnSignature(
+    fn: RsFunction,
+    boundsLifetimes: List<ReferenceLifetime>
+): Pair<List<ReferenceLifetime>, List<ReferenceLifetime>> {
+    // these will collect all the lifetimes for references in arg/return types
+    val inputCollector = LifetimesCollector()
+    val outputCollector = LifetimesCollector()
+
+    // extract lifetimes in input argument types
+    fn.valueParameterList?.let {
+        it.selfParameter?.accept(inputCollector)
+        it.valueParameterList.forEach { it.typeReference?.accept(inputCollector) }
+    }
+
+    // extract lifetimes in output type
+    fn.retType?.typeReference?.accept(outputCollector)
+
+    val inputLifetimes = inputCollector.lifetimes + boundsLifetimes
+    val outputLifetimes = outputCollector.lifetimes
+    return Pair(inputLifetimes, outputLifetimes)
+}
+
+private fun allowedLifetimesFrom(lifetimeParameters: List<RsLifetimeParameter>): Set<ReferenceLifetime> {
+    val allowedLifetimes = HashSet<ReferenceLifetime>()
+    lifetimeParameters
+        .filter { it.bounds.isEmpty() }
+        .mapNotNull { it.name }
+        .map { ReferenceLifetime.Named(it) }
+        .toCollection(allowedLifetimes)
+    allowedLifetimes.add(ReferenceLifetime.Unnamed)
+    allowedLifetimes.add(ReferenceLifetime.Static)
+    return allowedLifetimes
+}
+
+/**
+ * Are any lifetimes mentioned in the `where` clause? If yes, we don't try to reason about elision.
+ */
+private fun hasWhereLifetimes(whereClause: RsWhereClause?): Boolean {
+    whereClause ?: return false
+    for (predicate in whereClause.wherePredList) {
+        if (predicate.lifetime != null) return true
+
+        // a predicate like F: Trait or F: for<'a> Trait<'a>
+
+        // check the type F, it may not contain LT refs
+        val collector = LifetimesCollector()
+        predicate.typeReference?.accept(collector)
+        if (collector.lifetimes.isNotEmpty()) return true
+
+        // if the bounds define new lifetimes, they are fine to occur
+        val boundLifetimeParams = predicate.forLifetimes?.lifetimeParameterList.orEmpty()
+        val allowedLifetimes = allowedLifetimesFrom(boundLifetimeParams)
+        // now walk the bounds
+        predicate.typeParamBounds?.polyboundList?.map { it.bound }?.forEach { it.accept(collector) }
+        // and check that all lifetimes are allowed
+        if (!allowedLifetimes.containsAll(collector.lifetimes)) return false
+    }
+    return false
+}
+
+private fun checkLifetimesUsedInBody(body: RsBlock?): Boolean {
+    body ?: return false
+    val checker = BodyLifetimeChecker()
+    body.accept(checker)
+    return checker.lifetimesUsedInBody
+}
+
+private fun LifetimeName.toReferenceLifetime(): ReferenceLifetime =
+    when (this) {
+        is LifetimeName.Parameter -> ReferenceLifetime.Named(name)
```

Comment:
Maybe import `org.rust.lang.core.types.LifetimeName.*` and use just `Parameter`, `Static` and so on?